doc/guide.md

```diff
@@ -0,0 +1,241 @@
+| Header File Name | `libtock/[category]/internal/[name]_internal.h` |
+
+### Upcalls
+
+Each supported upcall must have its own function.
+
+The signature is:
+
+```c
+returncode_t set_upcall(subscribe_upcall callback, void* callback_args);
+```
+
+If only one upcall is supported, the function name must be `[name]_set_upcall`.
+
+If more than one upcall is supported, the function names must start with
+`[name]_set_upcall_` followed by a description of what the upcall is used for.
+
+
+#### Example:
+
+```c
+returncode_t [name]_set_upcall(subscribe_upcall callback, void* callback_args) {
+  subscribe_return_t sval = subscribe(DRIVER_NUM_[NAME], 0, callback, callback_args);
+  return tock_subscribe_return_to_returncode(sval);
+}
+```
+
+### Allows
+
+Each supported allow (readonly, readwrite, userspace_read) must have its own
+function.
+
+The signature is:
+
+- Read-Only Allow:
+    ```c
+    int set_roallow(const uint8_t* buffer, uint32_t len);
+    ```
+
+- Read-Write Allow:
+    ```c
+    int set_rwallow(uint8_t* buffer, uint32_t len);
+    ```
+
+- Userspace Read Allow:
+    ```c
+    int set_urallow(uint8_t* buffer, uint32_t len);
+    ```
+
+If only one allow is supported, the function name must be
+`[name]_set_[type]allow`.
+
+If more than one allow is supported, the function names must start with
+`[name]_set_[type]allow_` followed by a description of what the allow is used
+for.
+
+#### Example:
+
+```c
+returncode_t [name]_set_roallow_[desc](const uint8_t* buffer, uint32_t len) {
+  allow_ro_return_t aval = allow_readonly(DRIVER_NUM_[NAME], 0, (void*) buffer, len);
+  return tock_allow_ro_return_to_returncode(aval);
+}
+```
+
+```c
+returncode_t [name]_set_rwallow_[desc](uint8_t* buffer, uint32_t len) {
+  allow_rw_return_t aval = allow_readwrite(DRIVER_NUM_[NAME], 0, (void*) buffer, len);
+  return tock_allow_rw_return_to_returncode(aval);
+}
+```
+
+```c
+returncode_t [name]_set_urallow_[desc](uint8_t* buffer, uint32_t len) {
+  allow_userspace_r_return_t aval = allow_userspace_read(DRIVER_NUM_[NAME], 0, (void*) buffer, len);
+  return tock_allow_userspace_r_return_to_returncode(aval);
+}
+```
+
+### Commands
+
+Each supported command must have its own function.
+
+The signature is:
+
+```c
+returncode_t command(<arguments>);
+```
+
+For every command, the function names must start with `[name]_command_` followed
+by a description of what the command is used for.
+
+### Example:
+
+```c
+returncode_t [name]_command_[desc](void) {
+  syscall_return_t cval = command(DRIVER_NUM_[NAME], 1, 0, 0);
+  return tock_command_return_novalue_to_returncode(cval);
+}
+```
+
+
+### Exists
+
+There must be a function to check for syscall driver existence.
+
+Signature:
+
+```
+bool [name]_exists(void);
+```
+
+Example:
+
+```c
+bool [name]_exists(void) {
+  return driver_exists(DRIVER_NUM_[NAME]);
+}
+```
+
+
+## Asynchronous APIs
+
+All common system call operations should have asynchronous versions.
```

Comment:
I think it might be helpful to describe more what constitutes a "common system call operation".